### ISSUE TITLE
Lint integer variable fallback to i32

### ISSUE THREAD

**Issue by tamuhey** [rust-lang/rust-clippy#6634]
From https://stackoverflow.com/questions/65867874/how-to-deny-int-variable-without-explicit-type-annotation-in-rust?noredirect=1#comment116458771_65867874

Related: https://github.com/rust-lang/rfcs/blob/master/text/0212-restore-int-fallback.md

### What it does

Reports integer variable fallback to i32, e.g. 

```rust
let mut counter = 0; # Warning, because the type of `counter` is ambiguous and fallback to `i32`
let mut counter: u64 = 0; # OK
let mut counter = 0usize # OK
```
 
The below code should also be accepted because the variable is inferred though it is not annotated:

```rust
fn f(x: usize);
let mut counter = 0; # OK, because the variable `counter` is inferred as `usize` later.
f(counter)
```

### Categories (optional)

- Kind: *See <https://github.com/rust-lang/rust-clippy/blob/master/README.md#clippy> for list of lint kinds*

maybe `pedantic`

*What is the advantage of the recommended code over the original code*

I want this lint rule because it can cause overflow unintendedly:

```rust
let mut counter = 0;
for _ in 0..1_000_000_000_000usize {
    counter += 1;  # overflow!
    ...
}
```

### Drawbacks

May feel tedious

### Example

```rust
let counter = 0;
```

Could be written as:

```rust
let counter = 0usize;
let counter: usize = 0;
// or make `couter` to be inferred as `usize`
f(counter); // f: (usize) -> (); for example.
```

**Comment by Y-Nak** [comment]
This issue seems duplicate of #6064.
